heudiconv/dicoms.py

```diff
@@ -6,25 +6,154 @@
+        # For demographics to populate BIDS participants.tsv
+        patient_age=dcminfo.get('PatientAge'),
+        patient_sex=dcminfo.get('PatientSex'),
+        date=dcminfo.get('AcquisitionDate'),
+        series_uid=dcminfo.get('SeriesInstanceUID')
+    )
+    return seqinfo
+
 
-def group_dicoms_into_seqinfos(files, file_filter, dcmfilter, grouping):
+def validate_dicom(fl, dcmfilter):
+    """
+    Parse DICOM attributes. Returns None if not valid.
+    """
+    mw = dw.wrapper_from_file(fl, force=True, stop_before_pixels=True)
+    # clean series signature
+    for sig in ('iop', 'ICE_Dims', 'SequenceName'):
+        try:
+            del mw.series_signature[sig]
+        except Exception:
+            pass
+    # Workaround for protocol name in private siemens csa header
+    if not getattr(mw.dcm_data, 'ProtocolName', '').strip():
+        mw.dcm_data.ProtocolName = parse_private_csa_header(
+            mw.dcm_data, 'ProtocolName', 'tProtocolName'
+        ) if mw.is_csa else ''
+    try:
+        series_id = (
+            int(mw.dcm_data.SeriesNumber), mw.dcm_data.ProtocolName
+        )
+    except AttributeError as e:
+        lgr.warning(
+            'Ignoring %s since not quite a "normal" DICOM: %s', fl, e
+        )
+        return None
+    if dcmfilter is not None and dcmfilter(mw.dcm_data):
+        lgr.warning("Ignoring %s because of DICOM filter", fl)
+        return None
+    if mw.dcm_data[0x0008, 0x0016].repval in (
+        'Raw Data Storage',
+        'GrayscaleSoftcopyPresentationStateStorage'
+    ):
+        return None
+    try:
+        file_studyUID = mw.dcm_data.StudyInstanceUID
+    except AttributeError:
+        lgr.info("File {} is missing any StudyInstanceUID".format(fl))
+        file_studyUID = None
+    return mw, series_id, file_studyUID
+
+
+def group_dicoms_into_seqinfos(files, grouping, file_filter=None,
+                               dcmfilter=None, flatten=False,
+                               custom_grouping=None):
     """Process list of dicoms and return seqinfo and file group
     `seqinfo` contains per-sequence extract of fields from DICOMs which
     will be later provided into heuristics to decide on filenames
     Parameters
     ----------
     files : list of str
       List of files to consider
+    grouping : str
+      Possible groupings: studyUID, accession_number, all, custom
     file_filter : callable, optional
       Applied to each item of filenames. Should return True if file needs to be
       kept, False otherwise.
     dcmfilter : callable, optional
       If called on dcm_data and returns True, it is used to set series_id
-    grouping : {'studyUID', 'accession_number', None}, optional
-        what to group by: studyUID or accession_number
+    flatten : bool, optional
+      Creates a flattened `seqinfo` with corresponding DICOM files. True when
+      invoked with `dicom_dir_template`.
+    custom_grouping: variable or method, optional
```

Comment:
so default is `None`. Not clear to me (either from docstring here or from glancing over the code) what would happen if I specify `grouping='custom'` but do not specify `custom_grouping`.  Please clarify here if there is an intended behavior. If not - add a check that `custom_grouping` is not None or smth alike (e.g. if it is not callable - it is a string)